hashashin/main.py

```diff
@@ -0,0 +1,227 @@
+import glob
+import os
+from abc import ABC
+from dataclasses import dataclass
+from enum import Enum
+from pathlib import Path
+from typing import Collection, Iterable, Optional, Union
+import argparse
+
+import magic
+from tqdm import tqdm
+
+from hashashin.classes import BinarySignature
+from hashashin.db import (
+    BinarySignatureRepository,
+    FunctionFeatureRepository,
+    SQLAlchemyBinarySignatureRepository,
+    SQLAlchemyFunctionFeatureRepository,
+)
+from hashashin.feature_extractors import BinjaFeatureExtractor, FeatureExtractor
+from hashashin.utils import get_binaries
+import logging
+
+
+logger = logging.getLogger(os.path.basename(__file__))
+
+
+class Task(Enum):
+    HASH = "hash"
+    MATCH = "match"
+
+
+@dataclass
+class HashashinApplicationContext:
+    extractor: FeatureExtractor
+    feature_repo: FunctionFeatureRepository
+    binary_repo: BinarySignatureRepository
+    target_path: Optional[Path]
+    save_to_db: Optional[bool]
+    sig_match_threshold: float = 0.5
+    # feat_match_threshold: int = 0
+    progress: bool = False
+
+
+@dataclass
+class HashashinFactoryContext:
+    app_context: HashashinApplicationContext
+    task: Task
+
+
+class HashashinApplication(ABC):
+    def __init__(self, context: HashashinApplicationContext):
+        self.context = context
+
+    def run(self):
+        raise NotImplementedError
+
+    def saveToDB(
+        self,
+        signatures: Union[Collection[BinarySignature], BinarySignature],
+    ):
+        if isinstance(signatures, BinarySignature):
+            signatures = [signatures]
+        for sig in signatures:
+            binary = self.context.binary_repo.store_signature(sig)
+            for feat in sig.functionFeatureList:
+                feat.binary_id = binary.id
+                self.context.feature_repo.store_feature(feat)
```

Comment:
Seems like a dangerous way to interact with the database, is there a better way to do this on the `Repository` class level?

Reply:
This might be indication that there should really only be one `Repository` to interface with which stores these things in separate tables under the hood but doesn't affect how the user queries it.

Reply:
Addressed with 7f52633c3cdcff09739c8127f0407cb999da2b44.